index.js

```diff
@@ -22,7 +22,11 @@ function fastifySwagger (fastify, opts, next) {
   }
 }
 
-module.exports = fp(fastifySwagger, {
+const plugin = fp(fastifySwagger, {
   fastify: '>=3.x',
   name: 'fastify-swagger'
 })
+
+plugin.default = fastifySwagger
```

Comment:
I think we should do this within `fastify-plugin` so all fastify plugins automatically work instead.

Reply:
Yes, I agree with you. I am going to make a PR on `fastify-plugin` later since I am still figuring out the correct types that do not break anything. We will also need a PR to Fastify itself, for the `.regsiter()` method typings. Roughly something like this:
```
import { FastifyInstance } from './instance'
import { RawServerBase, RawRequestDefaultExpression, RawReplyDefaultExpression, RawServerDefault } from './utils'

/**
 * FastifyPluginCallback
 *
 * Fastify allows the user to extend its functionalities with plugins. A plugin can be a set of routes, a server decorator or whatever. To activate plugins, use the `fastify.register()` method.
 */

interface FastifyPluginCallback<Options extends FastifyPluginOptions = {}, Server extends RawServerBase = RawServerDefault> {
  (
    instance: FastifyInstance<Server, RawRequestDefaultExpression<Server>, RawReplyDefaultExpression<Server>>,
    opts: Options,
    next: (err?: Error) => void
  ): void
  default: (
    instance: FastifyInstance<Server, RawRequestDefaultExpression<Server>, RawReplyDefaultExpression<Server>>,
    opts: Options,
    next: (err?: Error) => void
  ) => void
}

/**
 * FastifyPluginAsync
 *
 * Fastify allows the user to extend its functionalities with plugins. A plugin can be a set of routes, a server decorator or whatever. To activate plugins, use the `fastify.register()` method.
 */
interface FastifyPluginAsync<Options extends FastifyPluginOptions = {}, Server extends RawServerBase = RawServerDefault> {
  (
    instance: FastifyInstance<Server, RawRequestDefaultExpression<Server>, RawReplyDefaultExpression<Server>>,
    opts: Options
  ): Promise<void>
  default:  (
    instance: FastifyInstance<Server, RawRequestDefaultExpression<Server>, RawReplyDefaultExpression<Server>>,
    opts: Options
  ) => Promise<void>
}

/**
 * Generic plugin type.
 * @deprecated union type doesn't work well with type inference in TS and is therefore deprecated in favor of explicit types. See FastifyRegister.
 */
export type FastifyPlugin<Options extends FastifyPluginOptions = {}> = FastifyPluginCallback<Options> | FastifyPluginAsync<Options>

export interface FastifyPluginOptions {
  [key: string]: any;
}

```

Reply:
However I think that we have to encourage all plugin developers to always export also the default prop in order to be sure that even if FP is not used, the plugin will work in every scenario.
What do you think about this? Do you think it is ok to put a note in the docs (not only in the TS section but also in the JS one)?

Reply:
Let's do the changes to fastify itself, to fastify-plugin, and the note docs.

Reply:
Ok just please check the quirks at https://github.com/fastify/fastify/issues/2403#issuecomment-659470548

I really want to make sure that we all agree and are on the same page, since I finally got the big picture. :)

Reply:
any update @fox1t?

Reply:
Hi, I just published a document that recaps TS/ES modules. We can use it to decide what we want to achieve.